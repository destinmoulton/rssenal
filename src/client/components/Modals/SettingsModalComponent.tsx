import * as React from "react";

import { Button, Header, Icon, Modal, Popup, Radio } from "semantic-ui-react";

import { ISetting, TSettings } from "../../interfaces";

export interface ISettingsModalMapDispatch {
    settingChange: (setting_key: string, setting_value: any) => void;
    refreshAllFeeds: () => void;
}

export interface ISettingsModalMapState {
    settings: TSettings;
}

type TAllProps = ISettingsModalMapState & ISettingsModalMapDispatch;

interface ISettingsModalState {
    isModalOpen: boolean;
}

class SettingsModalComponent extends React.Component<
    TAllProps,
    ISettingsModalState
> {
    state = {
        isModalOpen: false
    };

    componentDidUpdate() {
        console.log(this.props);
    }

    _handleClickOpenModal = () => {
        this.setState({
            isModalOpen: true
        });
    };

    _handleClickCloseModal = () => {
        this.setState({
            isModalOpen: false
        });
    };

    _buildSettings() {
        const { settings } = this.props;

        return settings
            .map((setting, key) => {
                if (setting.type === "toggle") {
                    return this._buildRadioToggle(key, setting);
                }
            })
            .toArray();
    }

    _buildRadioToggle(setting_key: string, setting: ISetting) {
        return (
            <div className="rss-settings-toggle-box" key={setting_key}>
                <Radio
                    checked={setting.value}
                    toggle
                    label={setting.name}
                    onChange={this._handleChangeSetting.bind(
                        this,
                        setting_key,
                        !setting.value,
                        setting
                    )}
                />
            </div>
        );
    }

    _handleChangeSetting(
        setting_key: string,
        setting_value: any,
        setting: ISetting
    ) {
        this.props.settingChange(setting_key, setting_value);

        if (setting.refresh_entries_on_change) {
            this.props.refreshAllFeeds();
        }
    }

    render() {
        const { isModalOpen } = this.state;
        const content = this._buildSettings();
        return (
            <span>
                <Popup
                    trigger={
                        <Icon
                            name="setting"
                            link={true}
                            size="large"
                            color="blue"
                            onClick={this._handleClickOpenModal}
                        />
                    }
                    content="Settings"
                />
                <Modal
                    open={isModalOpen}
                    onClose={this._handleClickCloseModal}
                    size="tiny"
                    className="rss-modal"
                >
                    <Header icon="setting" content="Settings" />
                    <Modal.Content>{content}</Modal.Content>
                    <Modal.Actions>
                        <Button
                            color="green"
                            onClick={this._handleClickCloseModal}
                            inverted
                        >
                            <Icon name="checkmark" /> Done
                        </Button>
                    </Modal.Actions>
                </Modal>
            </span>
        );
    }
}

export default SettingsModalComponent;
