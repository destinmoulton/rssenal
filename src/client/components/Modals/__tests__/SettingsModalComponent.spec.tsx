import * as React from "react";

import { shallow } from "enzyme";

import SettingsModalComponent from "../SettingsModalComponent";

import settings from "../../../../../test/data/settings";

describe("<SettingsModalComponent />", () => {
    it("renders and matches the snapshot", () => {
        const wrapper = shallow(
            <SettingsModalComponent
                settings={settings}
                settingChange={jest.fn()}
                refreshAllFeeds={jest.fn()}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });
});
