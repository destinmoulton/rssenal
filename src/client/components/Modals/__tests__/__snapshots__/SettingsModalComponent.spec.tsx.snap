// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsModalComponent /> renders and matches the snapshot 1`] = `
<span>
  <Popup
    content="Settings"
    keepInViewPort={true}
    on="hover"
    position="top left"
    trigger={
      <Icon
        as="i"
        color="blue"
        link={true}
        name="setting"
        onClick={[Function]}
        size="large"
      />
    }
  />
  <Modal
    centered={true}
    className="rss-modal"
    closeOnDimmerClick={true}
    closeOnDocumentClick={false}
    dimmer={true}
    eventPool="Modal"
    onClose={[Function]}
    open={false}
    size="tiny"
  >
    <Header
      content="Settings"
      icon="setting"
    />
    <ModalContent>
      <div
        className="rss-settings-toggle-box"
        key="show_images"
      >
        <Radio
          checked={true}
          label="Show Images"
          onChange={[Function]}
          toggle={true}
          type="radio"
        />
      </div>
      <div
        className="rss-settings-toggle-box"
        key="show_entries_has_read"
      >
        <Radio
          checked={false}
          label="Show Entries Already Read"
          onChange={[Function]}
          toggle={true}
          type="radio"
        />
      </div>
    </ModalContent>
    <ModalActions>
      <Button
        as="button"
        color="green"
        inverted={true}
        onClick={[Function]}
        role="button"
      >
        <Icon
          as="i"
          name="checkmark"
        />
         Done
      </Button>
    </ModalActions>
  </Modal>
</span>
`;
